File: Solución ejercicio Práctico Mayo 2025/backend/service/AdService.java
@Service
public class AdService {

	@Autowired
	private AdRepository adRepository;

    @Autowired
	private AdMapper mapper;

	public Collection<AdDTO> getAds(boolean rented) {
		return toDTOs(adRepository.findAllByRentedOrderByPrice(rented));
	}

	public AdDTO getAd(long id) {
		return toDTO(adRepository.findById(id).orElseThrow());
	}

	public AdDTO createAd(AdDTO adDTO) {

		if (adRepository.findByTitle(adDTO.title()).isPresent()) {
			return null;
		}

		Ad ad = toDomain(adDTO);

		adRepository.save(adRepository);

		return toDTO(ad);
	}

	public AdDTO replaceAd(long id, AdDTO updatedAdDTO) {

		Ad oldAd = adRepository.findById(id).orElseThrow();
		Ad updatedAd = toDomain(updatedAdDTO);
		updatedAd.setId(id);

		adRepository.save(updatedAd);

		return toDTO(updatedAd);
	}

	public AdDTO deleteAd(long id) {

		Ad ad = adRepository.findById(id).orElseThrow();

		AdDTO adDTO = toDTO(ad);

		adRepository.deleteById(id);

		return adDTO;
	}

	private AdDTO toDTO(Ad ad) {
		return mapper.toDTO(ad);
	}

	private Ad toDomain(AdDTO adDTO) {
		return mapper.toDomain(adDTO);
	}

	private Collection<AdDTO> toDTOs(Collection<Ad> ads) {
		return mapper.toDTOs(ads);
	}
}

File: Solución ejercicio Práctico Mayo 2025/frontend/service/ad.service.ts
const BASE_URL = "/api/ads/";

@Injectable({ providedIn: "root" })
export class AdService {
  constructor(private httpClient: HttpClient) {}

  public getAds(rented: boolean): Observable<AdDTO[]> {
    let params = new HttpParams();
    params = params.append("rented", rented);

    return this.httpClient.get(BASE_URL, { params: params }) as Observable<
      AdDTO[]
    >;
  }

  public getAd(id: number): Observable<AdDTO> {
    return this.httpClient.get(BASE_URL + id) as Observable<AdDTO>;
  }

  public createAd(ad: AdDTO): Observable<AdDTO> {
    return this.httpClient.post(BASE_URL, ad) as Observable<AdDTO>;
  }

  public replaceAd(ad: AdDTO): Observable<AdDTO> {
    return this.httpClient.put(BASE_URL + ad.id, ad) as Observable<AdDTO>;
  }

  public deleteAd(id: number): Observable<AdDTO> {
    return this.httpClient.delete(BASE_URL + id) as Observable<AdDTO>;
  }
}
